src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,323 @@ namespace System.Net.Sockets
-                                var eng = (SocketAsyncEngine)o;
-                                for (;;)
-                                {
-                                    try
-                                    {
-                                        eng.EventLoop();
-                                    }
-                                    catch (Exception e)
-                                    {
-                                        Debug.Fail(string.Format("Exception thrown from event loop: {0}", e.Message));
-                                    }
-                                }
-                            }, engine, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
-
-                            Volatile.Write(ref _engine, engine);
-                        }
+        private IntPtr AllocateHandle(SocketAsyncContext context)
+        {
+            Debug.Assert(Monitor.IsEntered(s_lock));
+            Debug.Assert(!IsFull);
+
+            IntPtr handle = _nextHandle;
+            _handleToContextMap.Add(handle, context);
+
+            _nextHandle = IntPtr.Add(_nextHandle, 1);
+            _outstandingHandles = IntPtr.Add(_outstandingHandles, 1);
+
+            if (IsFull)
+            {
+                // We'll need to create a new event port for the next handle.
+                s_currentEngine = null;
+            }
+
+            Debug.Assert(handle != ShutdownHandle);
+            return handle;
+        }
+
+        private void FreeHandle(IntPtr handle)
+        {
+            Debug.Assert(handle != ShutdownHandle);
+
+            bool shutdownNeeded = false;
+
+            lock (s_lock)
+            {
+                if (_handleToContextMap.Remove(handle))
+                {
+                    _outstandingHandles = IntPtr.Subtract(_outstandingHandles, 1);
+                    Debug.Assert(_outstandingHandles.ToInt64() >= 0);
+
+                    //
+                    // If we've allocated all possible handles for this instance, and freed them all, then 
+                    // we don't need the event loop any more, and can reclaim resources.
+                    //
+                    if (IsFull && _outstandingHandles == IntPtr.Zero)
+                    {
+                        shutdownNeeded = true;
                     }
                 }
+            }
 
-                return _engine;
+            //
+            // Signal shutdown outside of the lock to reduce contention.  
+            //
+            if (shutdownNeeded)
+            {
+                Shutdown();
             }
         }
 
-        private SocketAsyncEngine(int port, Interop.Sys.SocketEvent* buffer)
+        private SocketAsyncContext GetContextFromHandle(IntPtr handle)
         {
-            _port = port;
-            _buffer = buffer;
+            Debug.Assert(handle != ShutdownHandle);
+            Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64());
+            lock (s_lock)
```

Comment:
It's a (small-ish) concern of mine as well.  Agreed that we should run perf tests to find bottlenecks. :)